File: Algorithm and Data Structure/Baekjoon/16235번 나무 재테크/boj16235_rebas.py
from sys import stdin
from collections import deque
input = stdin.readline

ans = 0
dx, dy = (-1, -1, -1, 0, 0, 1, 1, 1), (-1, 0, 1, -1, 1, -1, 0, 1)
N, M, K = map(int, input().split())
a = [list(map(int, input().split())) for _ in range(N)]
tree = [[deque() for _ in range(N)] for _ in range(N)]
nutrient = [[5 for _ in range(N)] for _ in range(N)]
for _ in range(M):
    x, y, z = map(int, input().split())
    tree[x-1][y-1].append(z)
    ans += 1

def spring_summer():
    global ans
    for i in range(N):
        for j in range(N):
            for k in range(len(tree[i][j])):
                if nutrient[i][j] >= tree[i][j][k]:
                    nutrient[i][j] -= tree[i][j][k]
                    tree[i][j][k] += 1
                else:
                    while k < len(tree[i][j]):
                        nutrient[i][j] += (tree[i][j].pop()//2)
                        ans -= 1
                    break

def fall_winter():
    global ans
    for i in range(N):
        for j in range(N):
            for k in range(len(tree[i][j])):
                if tree[i][j][k]%5 == 0:
                    for t in range(8):
                        x, y = i+dx[t], j+dy[t]
                        if x < 0 or x >= N or y < 0 or y >= N:
                            continue
                        tree[x][y].appendleft(1)
                        ans += 1
            nutrient[i][j] += a[i][j]

def solve():
    for _ in range(K):
        spring_summer()
        fall_winter()
    print(ans)

solve()

File: Algorithm and Data Structure/Baekjoon/16000번대/16200번대/16235번 나무 재테크/boj16235_rebas.cc
#include <cstdio>
#include <deque>
using namespace std;

int N, M, K, ans;
int a[10][10];
int nutrient[10][10];
deque<int> tree[10][10];
const int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
const int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};

void spring_summer() {
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            int &nt = nutrient[i][j];
            deque<int> &tr = tree[i][j];
            for (int k=0; k<(int)tr.size(); k++) {
                if (nt >= tr[k]) { // Spring
                    nt -= tr[k];
                    tr[k]++;
                } else { // Summer
                    while (k < (int)tr.size()) {
                        nt += (tr.back()/2);
                        tr.pop_back();
                        ans--;
                    }
                    break;
                }
            }
        }
    }
}

void fall_winter() {
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            for (int k=0; k<(int)tree[i][j].size(); k++) {
                if (tree[i][j][k]%5 == 0) { // Fall
                    for (int t=0; t<8; t++) {
                        int x = i+dx[t], y = j+dy[t];
                        if (x < 0 || x >= N || y < 0 || y >= N) continue;
                        tree[x][y].push_front(1);
                        ans++;
                    }
                }
            }
            nutrient[i][j] += a[i][j]; // Winter
        }
    }
}

void solve() {
    while (K--) {
        spring_summer();
        fall_winter();
    }
    printf("%d\n", ans);
}

int main() {
    scanf("%d %d %d", &N, &M, &K);
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            scanf("%d", &a[i][j]);
            nutrient[i][j] = 5;
        }
    }
    for (int i=0; i<M; i++) {
        int x, y, z;
        scanf("%d %d %d", &x, &y, &z);
        tree[x-1][y-1].push_back(z);
        ans++;
    }
    solve();
    return 0;
}


